### ISSUE TITLE
Unable to use resource when moved & stashed inside component?

### ISSUE THREAD

**Issue by jcbhmr** [bytecodealliance/jco#337]
First the code context:

```wit
package jcbhmr:hello-world-rust-wasm-component-lib;

interface my-box-interface {
  resource my-box {
    constructor(value: u32);
    value: func() -> u32;
  }
  set-global: func(value: my-box);
  global-value: func() -> u32;
}

world example {
  export my-box-interface;
}

```

```rs
cargo_component_bindings::generate!();
use bindings::exports::jcbhmr::hello_world_rust_wasm_component_lib::*;
// use bindings::jcbhmr::hello_world_rust_wasm_component_lib::*;
use std::cell::RefCell;

pub struct Component;

thread_local!(static MY_BOX_GLOBAL: RefCell<my_box_interface::OwnMyBox> = RefCell::new(my_box_interface::OwnMyBox::new(MyBox(0))));
impl my_box_interface::Guest for Component {
    fn set_global(value: my_box_interface::OwnMyBox) {
        MY_BOX_GLOBAL.with(|global| {
            *global.borrow_mut() = value;
        });
    }
    fn global_value() -> u32 {
        MY_BOX_GLOBAL.with(|global| global.borrow().0)
    }
}

pub struct MyBox(u32);
impl my_box_interface::GuestMyBox for MyBox {
    fn new(value: u32) -> Self {
        Self(value)
    }
    fn value(&self) -> u32 {
        self.0
    }
}

```

```js
test("my box", () => {
  myBoxInterface.setGlobal(new myBox.MyBox(42));
  console.log(myBoxInterface.globalValue());
})
```

```js
test at file:/home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/tests/for-each.test.js:31:1
✖ my box (3.101ms)
  Error: Resource error: Invalid handle 1055504
      at trampoline7 (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/target/js/test/hello_world_rust_wasm_component_lib.js:1529:11)
      at _ZN35hello_world_rust_wasm_component_lib8bindings7exports6jcbhmr35hello_world_rust_wasm_component_lib6my_box98_$LT$impl$u20$wit_bindgen..RustResource$u20$for$u20$hello_world_rust_wasm_component_lib..MyBox$GT$3rep17h43571c84bd5b239cE (wasm://wasm/009c6b06:wasm-function[49]:0x272f)
      at _ZN74_$LT$wit_bindgen..Resource$LT$T$GT$$u20$as$u20$core..ops..deref..Deref$GT$5deref17h84f16cb406fa0a2eE (wasm://wasm/009c6b06:wasm-function[168]:0xa000)
      at _ZN181_$LT$hello_world_rust_wasm_component_lib..Component$u20$as$u20$hello_world_rust_wasm_component_lib..bindings..exports..jcbhmr..hello_world_rust_wasm_component_lib..my_box..Guest$GT$12global_value28_$u7b$$u7b$closure$u7d$$u7d$17hc44ed9686747c135E (wasm://wasm/009c6b06:wasm-function[133]:0x793a)
      at _ZN3std6thread5local17LocalKey$LT$T$GT$8try_with17h35aba8ea9a0d229bE (wasm://wasm/009c6b06:wasm-function[80]:0x3da6)
      at _ZN3std6thread5local17LocalKey$LT$T$GT$4with17hfdd362c94468220dE (wasm://wasm/009c6b06:wasm-function[79]:0x3bf9)
      at _ZN181_$LT$hello_world_rust_wasm_component_lib..Component$u20$as$u20$hello_world_rust_wasm_component_lib..bindings..exports..jcbhmr..hello_world_rust_wasm_component_lib..my_box..Guest$GT$12global_value17hcda847f1b372996bE (wasm://wasm/009c6b06:wasm-function[39]:0x24b9)
      at jcbhmr:hello-world-rust-wasm-component-lib/my-box#global-value (wasm://wasm/009c6b06:wasm-function[231]:0xbcb6)
      at Object.globalValue (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/target/js/test/hello_world_rust_wasm_component_lib.js:1801:89)
      at TestContext.<anonymous> (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/tests/for-each.test.js:33:21)
```

I want to be able to "move" (I think that's the right terminology) the resource _into the component_ and have the component own it. I assumed that this is what normally happened with the:

```wit
f: func(i-now-own-this: the-resource);
```

vs the borrowing approach which would leave the resource available to the host even after the component is done with it:

```wit
g: func(im-just-borrowing-this: borrow<the-resource>);
```

Thus, I assumed that I could stash ("move"?) the resource in the component's memory and use it at a later date. This doesn't appear to be the case. Am I misunderstanding? Is there a way to do what I want?

**Comment by jcbhmr** [comment]
Turns out this works in Rust Wasmtime! https://github.com/WebAssembly/component-model/issues/294#issuecomment-1891093627

**Comment by jcbhmr** [comment]
Here you go! https://github.com/jcbhmr/hello-world-rust-wasm-component-lib/tree/c030be0b86f575c83be23b20ba734b976160c72e
Note: the readme is completely wrong and outdated.

You can run it via:
```sh
make setup # If you want to auto-install the global tools
make build # To build the '.wasm' file
make bindings-js JCOFLAGS=--tracing # To generate and patch the JS bindings
make test-bindings-js # To run 'node --test'
make test-bindings-rs # To run 'cargo test'
```

This uses a slightly more complex example, but it is the original use-case that lead me to encounter this bug.

Specifically:

```wit
package jcbhmr:hello-world-rust-wasm-component-lib;

interface cbh {
  resource p-string {
    call: func(a: string);
  }
  resource r-string {
    call: func() -> string;
  }
}
interface cb {
  use cbh.{p-string as h-p-string, r-string as h-r-string};
  resource p-string {
    constructor(cb: h-p-string);
    call: func(a: string);
  }
  resource r-string {
    constructor(cb: h-r-string);
    call: func() -> string;
  }
}

interface hello-world-rust-wasm-component-lib {
  use cb.{p-string, r-string};
  set-cb: func(cb: p-string);
  run-cb-with-result-of: func(cb: borrow<r-string>);
}

world hello-world-rust-wasm-component-lib-world {
  import cbh;
  export cb;

  export hello-world-rust-wasm-component-lib;
}
```

Notice how the `interface cb` has a `p-string` resource which **takes in and acquires ownership of** the host's `cbh` `p-string` resource. I `set-cb()` to set the `p-string` callback. Then later I `run-cb-with-result-of()` another callback. And that `r-string` callback works! It's just that when attempting to run the original `p-string` callback that I passed to `set-cb()` earlier that it fails with the `Invalid handle 123456` error.

Here's the exact log from the test

```js
jcbhmr@PIG-2016:~/Documents/hello-world-rust-wasm-component-lib$ make test-bindings-js
node --experimental-default-type=module --test
f: [Function: f] { [length]: 1, [name]: 'f' }
g: PString { cb: [Function: f] { [length]: 1, [name]: 'f' } }
h: PString {
  [Symbol(resource)]: 1114136,
  [Symbol(nodejs.dispose)]: <ref *1> [Function: value] {
    [length]: 0,
    [name]: 'value',
    [prototype]: { [constructor]: [Circular *1] }
  }
}
[module="jcbhmr:hello-world-rust-wasm-component-lib/cb", function="[constructor]p-string"] call cb=[object Object]
[module="jcbhmr:hello-world-rust-wasm-component-lib/cb", function="[constructor]p-string"] return result=0
[module="jcbhmr:hello-world-rust-wasm-component-lib/hello-world-rust-wasm-component-lib", function="set-cb"] call cb=[object Object]
[module="jcbhmr:hello-world-rust-wasm-component-lib/hello-world-rust-wasm-component-lib", function="set-cb"] return 
h (after set): PString {
  [Symbol(resource)]: null,
  [Symbol(nodejs.dispose)]: [Function: emptyFunc] { [length]: 0, [name]: 'emptyFunc' }
}
i: [Function: i] { [length]: 0, [name]: 'i' }
j: RString { cb: [Function: i] { [length]: 0, [name]: 'i' } }
[module="jcbhmr:hello-world-rust-wasm-component-lib/cb", function="[constructor]r-string"] call cb=[object Object]
[module="jcbhmr:hello-world-rust-wasm-component-lib/cb", function="[constructor]r-string"] return result=0
[module="jcbhmr:hello-world-rust-wasm-component-lib/hello-world-rust-wasm-component-lib", function="run-cb-with-result-of"] call cb=[object Object]
k: RString {
  [Symbol(resource)]: 1114168,
  [Symbol(nodejs.dispose)]: <ref *1> [Function: value] {
    [length]: 0,
    [name]: 'value',
    [prototype]: { [constructor]: [Circular *1] }
  }
}
h (before run): PString {
  [Symbol(resource)]: null,
  [Symbol(nodejs.dispose)]: [Function: emptyFunc] { [length]: 0, [name]: 'emptyFunc' }
}
[module="jcbhmr:hello-world-rust-wasm-component-lib/cbh", function="[method]r-string.call"] call self=0
[module="jcbhmr:hello-world-rust-wasm-component-lib/cbh", function="[method]r-string.call"] return result="Hello world!"
✖ setCb() and runCbWithResultOf() (12.8984ms)
  Error: Resource error: Invalid handle 1114136
      at trampoline3 (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/target/js/hello_world_rust_wasm_component_lib.js:121:11)
      at _ZN35hello_world_rust_wasm_component_lib8bindings7exports6jcbhmr35hello_world_rust_wasm_component_lib2cb100_$LT$impl$u20$wit_bindgen..RustResource$u20$for$u20$hello_world_rust_wasm_component_lib..PString$GT$3rep17hd52d54f2fe72edaeE (wasm://wasm/007e6592:wasm-function[135]:0x6118)
      at _ZN74_$LT$wit_bindgen..Resource$LT$T$GT$$u20$as$u20$core..ops..deref..Deref$GT$5deref17h4e8a9ea27eb5d562E (wasm://wasm/007e6592:wasm-function[80]:0x494e)
      at _ZN210_$LT$hello_world_rust_wasm_component_lib..Component$u20$as$u20$hello_world_rust_wasm_component_lib..bindings..exports..jcbhmr..hello_world_rust_wasm_component_lib..hello_world_rust_wasm_component_lib..Guest$GT$21run_cb_with_result_of28_$u7b$$u7b$closure$u7d$$u7d$17h3ae60c6644376abaE (wasm://wasm/007e6592:wasm-function[40]:0x209d)
      at _ZN3std6thread5local17LocalKey$LT$T$GT$8try_with17ha892787d70a57407E (wasm://wasm/007e6592:wasm-function[18]:0xfc5)
      at _ZN3std6thread5local17LocalKey$LT$T$GT$4with17h1c7bc52bdf27aba2E (wasm://wasm/007e6592:wasm-function[17]:0xe1d)
      at _ZN210_$LT$hello_world_rust_wasm_component_lib..Component$u20$as$u20$hello_world_rust_wasm_component_lib..bindings..exports..jcbhmr..hello_world_rust_wasm_component_lib..hello_world_rust_wasm_component_lib..Guest$GT$21run_cb_with_result_of17h1f1b4822fa52197aE (wasm://wasm/007e6592:wasm-function[133]:0x6077)
      at jcbhmr:hello-world-rust-wasm-component-lib/hello-world-rust-wasm-component-lib#run-cb-with-result-of (wasm://wasm/007e6592:wasm-function[29]:0x182e)
      at Object.runCbWithResultOf (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/target/js/hello_world_rust_wasm_component_lib.js:260:115)
      at TestContext.<anonymous> (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/tests/hello-world.test.js:33:34)

ℹ tests 1
ℹ suites 0
ℹ pass 0
ℹ fail 1
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 258.551803

✖ failing tests:

test at file:/home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/tests/hello-world.test.js:12:1
✖ setCb() and runCbWithResultOf() (12.8984ms)
  Error: Resource error: Invalid handle 1114136
      at trampoline3 (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/target/js/hello_world_rust_wasm_component_lib.js:121:11)
      at _ZN35hello_world_rust_wasm_component_lib8bindings7exports6jcbhmr35hello_world_rust_wasm_component_lib2cb100_$LT$impl$u20$wit_bindgen..RustResource$u20$for$u20$hello_world_rust_wasm_component_lib..PString$GT$3rep17hd52d54f2fe72edaeE (wasm://wasm/007e6592:wasm-function[135]:0x6118)
      at _ZN74_$LT$wit_bindgen..Resource$LT$T$GT$$u20$as$u20$core..ops..deref..Deref$GT$5deref17h4e8a9ea27eb5d562E (wasm://wasm/007e6592:wasm-function[80]:0x494e)
      at _ZN210_$LT$hello_world_rust_wasm_component_lib..Component$u20$as$u20$hello_world_rust_wasm_component_lib..bindings..exports..jcbhmr..hello_world_rust_wasm_component_lib..hello_world_rust_wasm_component_lib..Guest$GT$21run_cb_with_result_of28_$u7b$$u7b$closure$u7d$$u7d$17h3ae60c6644376abaE (wasm://wasm/007e6592:wasm-function[40]:0x209d)
      at _ZN3std6thread5local17LocalKey$LT$T$GT$8try_with17ha892787d70a57407E (wasm://wasm/007e6592:wasm-function[18]:0xfc5)
      at _ZN3std6thread5local17LocalKey$LT$T$GT$4with17h1c7bc52bdf27aba2E (wasm://wasm/007e6592:wasm-function[17]:0xe1d)
      at _ZN210_$LT$hello_world_rust_wasm_component_lib..Component$u20$as$u20$hello_world_rust_wasm_component_lib..bindings..exports..jcbhmr..hello_world_rust_wasm_component_lib..hello_world_rust_wasm_component_lib..Guest$GT$21run_cb_with_result_of17h1f1b4822fa52197aE (wasm://wasm/007e6592:wasm-function[133]:0x6077)
      at jcbhmr:hello-world-rust-wasm-component-lib/hello-world-rust-wasm-component-lib#run-cb-with-result-of (wasm://wasm/007e6592:wasm-function[29]:0x182e)
      at Object.runCbWithResultOf (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/target/js/hello_world_rust_wasm_component_lib.js:260:115)
      at TestContext.<anonymous> (file:///home/jcbhmr/Documents/hello-world-rust-wasm-component-lib/tests/hello-world.test.js:33:34)
make: *** [Makefile:30: test-bindings-js] Error 1
```

Compare that to the Rust Wasmtime version:

```sh
jcbhmr@PIG-2016:~/Documents/hello-world-rust-wasm-component-lib$ make test-bindings-rs
cargo test -- --show-output
    Finished test [unoptimized + debuginfo] target(s) in 0.81s
     Running unittests src/lib.rs (target/debug/deps/hello_world_rust_wasm_component_lib-345ad28eb7e3e82e)

running 0 tests

successes:

successes:

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/hello_world_test.rs (target/debug/deps/hello_world_test-fd60d0e37335046f)

running 1 test
test it_works ... ok

successes:

---- it_works stdout ----
my_cb recieved: a=Hello from my_cb2!


successes:
    it_works

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.07s
```

**Comment by guybedford** [comment]
@jcbhmr I'm getting the error bindings file `/Users/gbedford/Projects/hello-world-rust-wasm-component-lib/target/bindings/hello-world-rust-wasm-component-lib/bindings.rs` does not exist. Can you just share the component Binary for me to try here?

**Comment by guybedford** [comment]
Thanks for the replication here, I've been able to track this down and can confirm it is a bug.

I also found a simple fix - in the generated bindings with the following diff:

```diff
function trampoline3(handle) {
-  const handleEntry = handleTable2.get(handle);
-  if (!handleEntry) {
-    throw new Error(`Resource error: Invalid handle ${handle}`);
-  }
-  return handleEntry.rep;
+  return handle;
}
```

I still haven't worked out exactly how this is reflecting in the bindgen model, will aim to get a fix up soon, but it might take a few days.

**Comment by guybedford** [comment]
I spent some time today debugging this case with @alexcrichton and we found the issue, here is a rough summary of the case:

* When lowering a borrow, it is possible to simplify the handle in the case where that borrow is a resource created by that component itself to the direct rep pointer in the underlying component.
* But when lowering an own handle to an own resource, it must still be lowered as a handle and not as its rep, as we are currently doing.
* As a result, the problem happens in the `setCb` function already, where we are passing in the address instead of the handle.
* This is then why `trampoline3` is being called to get the address from the own handle, but instead it already has the address.

The fix is therefore to ensure that in `setCb` when lowering the own component's own resource, we lower into a handle and not a rep.

I will work on a fix for this next Friday, after the preview2 release, as this does not currently affect WASI applications, and that is the primary priority at the moment.
